### ISSUE TITLE
Question is related to 16-bit image, having trouble to save 16-bit image in JPEG including compression in python environment? 

### ISSUE THREAD

**Issue by Yuvi-416** [OpenImageIO/oiio#2964]
Hello there,
I asked this question over here: https://github.com/conda-forge/openimageio-feedstock/issues/47
but they asked me to post here.
Therefore, I would like to ask for some help regarding my problem.

I was trying to load an array of the 16-bit image using this package but I am getting an error as:
```
TypeError: init(): incompatible constructor arguments. The following argument types are supported:
1. OpenImageIO.ImageBuf()
2. OpenImageIO.ImageBuf(arg0: str)
3. OpenImageIO.ImageBuf(arg0: str, arg1: int, arg2: int)
4. OpenImageIO.ImageBuf(arg0: OpenImageIO.ImageSpec)
5. OpenImageIO.ImageBuf(arg0: OpenImageIO.ImageSpec, arg1: bool)
6. OpenImageIO.ImageBuf(name: str, subimage: int, miplevel: int, config: OpenImageIO.ImageSpec)

```
I understood the error, but I didn't understand how can I pass the array using this library. If it can pass then can I do something like the below which I did for an 8-bit image using PIL library:

```
img_tiles_org = Image.fromarray(bf_tiles_org).convert("RGB")
instance_byte_str_buffer_org = BytesIO()
img_tiles_org.save(instance_byte_str_buffer_org, "JPEG", quality=80, icc_profile=img_tiles_org.info.get('icc_profile'), progressive=False)
t_org = instance_byte_str_buffer_org.getvalue()
```
Before I was using Pillow library to save image bytes in jpeg compression for the 8-bit image but now I have a 16-bit image therefore, I can't use the PIL library anymore.
I also tried OpenCV-python but it has a limitation, it can't save a jpeg image of 16-bit as I understood.
Therefore, I would like to know if there is any library or package which can help me to load the bytes of an image and its ICC profile too.
I really need some suggestion regarding it.
Thank you

**Comment by lgritz** [comment]
Can you post what you are doing to get that error? What calls to OIIO are you making?

**Comment by Yuvi-416** [comment]
I was trying to load the 16-bit array which I receive from bf_tiles_org.  I was searching for some function that behaves exactly like Image.fromarray() function for 8-bit image in OpenImageIO.

Code:
```
import OpenImageIO
import OpenImageIO as oiio
from OpenImageIO import ImageBuf, ImageSpec, ImageBufAlgo
import bioformats
import javabridge

reader = bioformats.ImageReader(input_path)

for Series in range(reader.rdr.getSeriesCount()):  # czi
    reader.rdr.setSeries(Series)
    width = reader.rdr.getSizeX()
    height = reader.rdr.getSizeY()
    encoded_framed_items_org = []
    for zstack in range(reader.rdr.getImageCount()):
        bf_tiles_org = reader.read(z=zstack, rescale=False)
        img = ImageBuf(bf_tiles_org)
        print(img)
```
Error:
```
Traceback (most recent call last):
  File "/home/yuvi/PycharmProjects/task-16bit/Temp16-1.py", line 28, in <module>
    img = ImageBuf(bf_tiles_org)
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. OpenImageIO.ImageBuf()
    2. OpenImageIO.ImageBuf(arg0: str)
    3. OpenImageIO.ImageBuf(arg0: str, arg1: int, arg2: int)
    4. OpenImageIO.ImageBuf(arg0: OpenImageIO.ImageSpec)
    5. OpenImageIO.ImageBuf(arg0: OpenImageIO.ImageSpec, arg1: bool)
    6. OpenImageIO.ImageBuf(name: str, subimage: int, miplevel: int, config: OpenImageIO.ImageSpec)

Invoked with: array([[[1929, 2120, 1624],
        [1851, 2237, 1578],
        [1865, 2250, 1466],
        ...,
        [2276, 2117, 1878],
        [2288, 2128, 1781],
        [2293, 2106, 1790]],
        .....,
```
These are the errors that I am getting currently while I pass the ImageBuf function.

**Comment by lgritz** [comment]
Right, there is currently not an ImageBuf constructor directly from an array. You'd need to do something like this:

    spec = ImageSpec(width, height, channels, 'uint16')
    img = ImageBuf(spec)
    img.set_pixels(oiio.ROI(), bf_tiles_org)  # copy your array into the image
    buf.write('out.tif')

**Comment by Yuvi-416** [comment]
Hi,
I think you mean to write: img.write('out.tif') ?
The generated output was black, bcoz of the size of generated output, I can't upload it here.
And I also see that Adobe deflate compression is used as a default.
Is there a way to encode this image by utilizing BytesIO...

**Comment by lgritz** [comment]
Yes, I did mean `img.write('out.tif')`

Was the output image at least the right size? Can you send me the resulting file via email?  (lg AT larrygritz.com)

Deflate compression is used by default for tiff files. You can change the compression with:

    spec = ImageSpec(width, height, channels, 'uint16')
    # At this stage, you can set all sorts of interesting things about the spec:
    spec.attribute("compression", "lzw")
    ...
    img = ImageBuf(spec)
    ...

I don't know much about BytesIO, but I don't think there's an easy way to do that.

**Comment by Yuvi-416** [comment]
Hello there,
I send you an email with both input and as well as an output image, the input is a little bit bigger.
I tried to pass jpeg compression but I recognize that it only use lzw and Adobe deflate, although I pass JPEG but it changes to Adobe deflate.

**Comment by lgritz** [comment]
Hi, those images are just fine. Looks ok to me. Looks like a mosaic of images stitched together. Some kind of microscope capture?

    Stats Min: 0 66 0 (of 65535)
    Stats Max: 4095 4095 4095 (of 65535)
    Stats Avg: 1833.09 1969.62 1625.99 (of 65535)
    Stats StdDev: 565.16 631.24 431.19 (of 65535)

So though a 16 bit image can have pixel values up to 65535, the largest value in this image is 4095. That matches what you originally quoted above (with one value printed as [1851, 2237, 1578]). It may be "dark", but if you turn up the intensity a bit (most viewers can do something like this), it's very clear.

So I'm primarily suspecting that whatever you are using to *display* the image is broken, or can't handle 16 bit TIFF files properly, or it's not scaling the intensity of the image correctly so you are just seeing it very dark.

**Comment by lgritz** [comment]
As for the JPEG compression, the problem is that JPEG compression is only defined for 8 bit images. So the fact that this is a 16 bit image is correctly realizing that you have made an impossible request and will therefore revert to the default lossless compression, which is zip in this case.

**Comment by Yuvi-416** [comment]
Thank you for your suggestion!
